Add j2me_enabled for CommCareBuild(default=False) and set it to true …

mkangia opened the pull request:
…when using artifacts zip through CommCareBuild.create_from_zip.

Add ability on builds edit menu to update a build's j2me_enabled config and also set it according to j2me_enabled versions when changing CommCare version.
Display appt message when attempting to download Files for Java Phone in case of non-j2me-enabled saved app builds

[Spec](https://github.com/dimagi/commcare-hq/issues/12347)
[Fogbugz Ticket](http://manage.dimagi.com/default.asp?231812)

mkangia committed:
Add j2me_enabled for CommCareBuild(default=False) and set it to true when using artifacts zip through CommCareBuild.create_from_zip.
Add ability on builds edit menu to update a build's j2me_enabled config and also set it according to j2me_enabled versions when changing CommCare version.
Display appt message when attempting to download Files for Java Phone in case of non-j2me-enabled saved app builds

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/releases.js:
```diff
@@ -174,6 +174,9 @@ hqDefine('app_manager/js/releases.js', function () {
 
+        self.reveal_java_download = function(){
+            return this.j2me_enabled();
+        }
```
Missing semicolon. (semi)

dimagimon commented on corehq/apps/builds/models.py:
```diff
@@ -132,6 +133,17 @@ def get_build(cls, version, build_number=None, latest=False):
         return cls.view('builds/all', include_docs=True, reduce=False)
 
+    @classmethod
+    @quickcache([], timeout=5 * 60) #This seems to be not working.
```
E261 at least two spaces before inline comment
E262 inline comment should start with '# '

czue replied:
what about it is not working?

mkangia replied:
I could not see it caching the value for 5 mins. It was querying couchDB for all builds. So was not sure if i implemented it correctly. Can you help understand this?

czue replied:
do you have caching enabled on your local? it looks like it should work to me.

dannyroberts commented on corehq/apps/builds/models.py:
```diff
@@ -32,7 +32,8 @@ class CommCareBuild(Document):
     version = SemanticVersionProperty()
     time = DateTimeProperty()
-    
+    j2me_enabled = BooleanProperty(default=False)
```
From https://github.com/dimagi/commcare-hq/issues/12347 (emphasis mine):
> *Have HQ detect J2ME eligibility* by whether or not the build has the right attachments and save that to j2me_enabled (which _defaults to true_ for backwards compatibility).
It's important that it default to `True` because currently (before introducing these changes) _all_ builds were j2me enabled, so if the value is absent (i.e. if the build was made before this feature was introduced) then it should be `True`.

dannyroberts commented:
I believe you said this isn't complete yet; looks like it's still missing the part that handles new POSTs from Jenkins or new imports for builds that are not J2ME enabled (and thus don't have an attachment.

That and one more comment, and + the minor lint ("dimagimon") errors. Overall this looks very much in the right direction!

czue commented on corehq/apps/app_manager/templates/app_manager/partials/releases_deploy_modal.html:
```diff
@@ -184,6 +184,9 @@ <h4 class="modal-title">{% trans "Deploying version" %} <span data-bind="text: v
                                 </li>
                             </ol>
+                            <p data-bind="if: !reveal_java_download()">
+                                Please use the latest supported J2me version to have download available for Java Phone.
```
this line should be marked for translation similar to the ones above

mkangia replied:
Right. Done.

czue commented on corehq/apps/builds/models.py:
```diff
@@ -132,6 +133,17 @@ def get_build(cls, version, build_number=None, latest=False):
+        for version_build in cls.all_builds():
+            if version_build.j2me_enabled:
+                j2me_enabled_builds.append(version_build.version)
+        return j2me_enabled_builds
```
a few other pythonic ways to write this:
list comprehension:
```python
return [build for build in cls.all_builds() if build.j2me_enabled
```
filter built in
```python
return filter(lambda build: build.j2me_enabled, cls.all_builds())
```

mkangia replied:
Awesome! Thanks Cory.

czue commented on corehq/apps/builds/models.py:
```diff
@@ -239,6 +251,16 @@ def get_menu(self, application_version=None):
+        for version_config in cls.fetch().menu:
+            if version_config.j2me_enabled:
+                j2me_enabled_configs.append(version_config.label)
+        return j2me_enabled_configs
```
(same as above, list comprehension would be better here so that you can return the labels)

mkangia replied:
Awesome again.

czue commented on corehq/apps/builds/templates/builds/edit_menu.html:
```diff
@@ -140,6 +152,7 @@ <h3 class="panel-title">Menu Options</h3>
                             <th>Version</th>
                             <th>Label</th>
                             <th>Superuser Only</th>
+                            <th>J2me Enabled</th>
```
these should also all be translated (looks like an existing issue for the others)

mkangia replied:
Alright. Adding for ones i can. Trying not to break anything.

dimagimon commented on corehq/apps/builds/views.py:
```diff
@@ -150,43 +151,46 @@ def import_build(request):
             '_id': build.get_id,
         }
-    })
+    })
```
W292 no newline at end of file

mkangia committed:
Use list comprehension and map for filtering for list, better method division and naming, Add latest j2me_enabled config in message for Java Download

mkangia committed:
Add missing translation call for text on edit_menu page for Editing CommCare builds

mkangia commented:
@dannyroberts @czue : Thank you for the review. Took care of them. Please have a look again.
Awaiting feedback.

czue commented:
build failure looks real:
```
======================================================================
ERROR: test_basic_app (corehq.apps.app_manager.tests.test_views.TestViews)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/test_views.py", line 152, in test_basic_app
    }, {'limit': 5})
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/test_views.py", line 122, in _json_content_from_get
    response = self.client.get(reverse(name, kwargs=kwargs), data)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 470, in get
    **extra)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 286, in get
    return self.generic('GET', path, secure=secure, **r)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 358, in generic
    return self.request(**r)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 440, in request
    six.reraise(*exc_info)
  File "/vendor/lib/python2.7/site-packages/django/core/handlers/base.py", line 111, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/vendor/lib/python2.7/site-packages/django/views/decorators/cache.py", line 38, in _cache_controlled
    response = viewfunc(request, *args, **kw)
  File "/mnt/commcare-hq-ro/corehq/apps/domain/decorators.py", line 101, in _inner
    return view_func(req, domain_name, *args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/views/releases.py", line 78, in paginate_releases
    j2me_enabled_configs = CommCareBuildConfig.j2me_enabled_configs_labels()
  File "/mnt/commcare-hq-ro/corehq/apps/builds/models.py", line 268, in j2me_enabled_configs_labels
    return map(lambda x: x.label, cls.j2me_enabled_configs())
  File "/mnt/commcare-hq-ro/corehq/util/quickcache.py", line 312, in inner
    return helper(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/quickcache.py", line 100, in __call__
    content = self.fn(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/builds/models.py", line 264, in j2me_enabled_configs
    return [build for build in cls.fetch().menu if build.j2me_enabled]
AttributeError: 'BuildMenuItem' object has no attribute 'j2me_enabled'
```

czue commented on corehq/apps/app_manager/templates/app_manager/partials/releases_deploy_modal.html:
```diff
@@ -184,6 +184,11 @@ <h4 class="modal-title">{% trans "Deploying version" %} <span data-bind="text: v
+                            <p data-bind="if: !reveal_java_download()">
+                                {% trans "This version is not compatible with Java Phones. Please create a new build with latest J2ME supported version, "%}
+                                {{ lastest_j2me_enabled_build }}
+                                {% trans "to install on Java Phone." %}
```
This should use the [blocktrans/with](https://docs.djangoproject.com/es/1.9/topics/i18n/translation/#std:templatetag-blocktrans) pattern in case other languages don't structure the sentence in the same way.

mkangia replied:
@czue : Using blocktrans without the with option since its a simple context variable.

mkangia committed:
Fix build by adding j2me_enabled field to BuildMenuItem as a placeholder for value to be set according to CommCare app for new builds added

mkangia committed:
Use transblock instead of multiple trans statements

czue commented:
no other feedback from me, but I think @dannyroberts should review again since he has much more context on the feature. also needs to be merged with master now. thanks for the speedy follow ups!

mkangia committed:
Set j2me_enabled on app for ajax response when adding a new build

mkangia committed:
Fix dimagimon error for new line at eof

dannyroberts committed:
Merge branch 'master' into mk-231812-j2me-config

Conflicts:
	corehq/apps/app_manager/views/releases.py

dannyroberts commented on corehq/apps/builds/models.py:
```diff
@@ -132,6 +140,16 @@ def get_build(cls, version, build_number=None, latest=False):
+        return [build for build in cls.all_builds() if build.j2me_enabled]
+
+    @classmethod
+    def j2me_enabled_builds_versions(cls):
```
Nitpick: I'd call this `j2me_enabled_build_versions` (not build**s** versions)

dannyroberts commented on corehq/apps/builds/models.py:
```diff
@@ -239,6 +258,19 @@ def get_menu(self, application_version=None):
+        return [build for build in cls.fetch().menu if build.j2me_enabled]
+
+    @classmethod
+    def j2me_enabled_configs_labels(cls):
```
Nitpick: I'd also call this `j2me_enabled_config_labels` (not config**s** labels)

dannyroberts committed:
minor naming changes

j2me_enabled_builds_versions -> j2me_enabled_build_versions
j2me_enabled_configs_labels -> j2me_enabled_config_labels

dannyroberts commented on corehq/apps/builds/templates/builds/edit_menu.html:
```diff
@@ -1,23 +1,35 @@
+            self.version.subscribe(function(newValue){
+                self.j2me_enabled(buildsMenu.j2me_enabled_versions.
+                                  includes(newValue)
+                                 );
```
This subscribe here is so when you change the version in the Version dropdown, the check mark also changes, right? Nicely done.

mkangia replied:
Bingo!
Thanks 👍

dannyroberts commented on corehq/apps/builds/templates/builds/edit_menu.html:
```diff
@@ -161,6 +179,9 @@ <h3 class="panel-title">Menu Options</h3>
                                 <input type="checkbox" value="superuser-only" data-bind="checked: superuser_only" />
                             </td>
+                            <td class='j2me_enabled'>
+                                <input type="checkbox" value="j2me_enabled" data-bind="checked: j2me_enabled" />
```
This is slightly off from the spec, but in an important way.

> **Include a J2ME Enabled column** in [the build menu UI](https://www.commcarehq.org/builds/edit_menu/); a simple column that contains a checkmark icon if the build is j2me-enabled and is blank otherwise. This is for debugging purposes and to potentially reduce support burden of any confusion that arises from the split.

So you want it to look something like

<img width="340" alt="screen shot 2016-07-14 at 7 58 13 pm" src="https://cloud.githubusercontent.com/assets/137212/16859505/608a6d8c-49fd-11e6-8bc0-bb418a29c14d.png">

with a checkmark icon that's not editable. (Since it's tied to the Version you pick and not independent, it'd be bad for the user to be able to edit it directly.) You can do that with something like

```html
                            <td class='j2me_enabled'>
                                <!--ko if: j2me_enabled-->
                                <i class="fa fa-check" data-bind="visible: j2me_enabled"></i>
                                <!--/ko-->
                            </td>
```

Looks like you already took care of setting to the correct value when you change the value of Version dropdown (where I commented above), so this might be the only change you have to make.

mkangia replied:
Oh my bad! i took checkmark as checkbox coz I thought in that way the intent of having a checkbox was to make it configurable where the version check was like a suggestion from builds and its files. I over thought it i guess. Would change it as you suggested.
Thanks 👍

dannyroberts commented:
This looks really well thought through! I left a comment above, but I think that's the last change before it's pretty much exactly what we were looking for.

Once you're done, I'd put the "awaiting QA" label on this PR and then put this branch on staging and ping @phillipm to test it. Once he's given the OK on it, I'm happy to have someone merge it (I'm offline tomorrow, so even if I'm not online).

czue commented:
@mkangia let me know @sravfeyn know if you need any help with putting the branch on staging.

mkangia committed:
Update the builds index to not allow user to change the j2me_enabled field.

mkangia commented:
@dannyroberts : Thanks for the feedback and the naming fixes. Made the final change. Requesting a final review.
Placed the branch name in staging.yml file in master. Would notify philip to test it once its deployed after having a look myself ensuring that its working smoothly.
Meanwhile, adding some test cases for changes done.

mkangia commented:
@phillipm : Please help test this. Its deployed on HQ-staging. I did a check for basics and seems to be working well. Awaiting feedback.

phillipm commented:
@mkangia tested it out on staging. With the param `build_number=5` I got `{"message": "u\"'5'\" not of type (<type 'int'>, <type 'long'>)", "error": 500}`. I'll offline with you as to how you can test this out

mkangia committed:
Parse the build_number to integer before creating the build from params supplied for creating builds without artefacts

mkangia commented:
@dannyroberts : Did a small change to [fix](https://github.com/dimagi/commcare-hq/pull/12411/commits/3c43d8f88407a13aaca7bd19276962d3d45e8436) issue while serving post request for creating builds, as reported by @phillipm. Please review it.
@phillipm :Post next deploy on staging you should be able to check this feature again.
Will check back for the travis build.

Thank you guys for your patience with this.

dannyroberts commented:
Looks great to me!

phillipm commented:
Tested this out on staging; it works well!

One thing that should probably be fixed:
If you post a build without artifacts and then bump the builder number and include artifacts, the edit_menu list doesn't register that there are now artifacts. This functionality will be helpful because it will allow us to be a little looser with our deploy process. That is, we can run the normal non-J2ME deploy and then overwrite the build on special releases where we want to do a J2ME deploy

mkangia committed:
Explicitly set j2me_enabled True when creating build with artifacts to ensure the value for j2me_enabled in case same version created with a higher build_number

mkangia commented:
@dannyroberts Can you guide how to fix the issue pointed out by @phillipm. From my understanding [this](https://github.com/dimagi/commcare-hq/pull/12411/commits/01dcef85d272485a4ed032ccb0d6de33c9139697) should be the possible fix but not sure. Can you share any insight or how this can be fixed/approached?

dannyroberts commented on corehq/apps/builds/models.py:
```diff
@@ -132,6 +141,16 @@ def get_build(cls, version, build_number=None, latest=False):
 
+    @classmethod
+    @quickcache([], timeout=5 * 60)
+    #This seems to be not working.
```
What seems not to be working?

dannyroberts commented:
My guess is that the problem @phillipm mentioned has more to do with `j2me_enabled_builds` being `quickcached` than with the actual value of `j2me_enabled` on the new build not being set correctly. (I mean to say, my guess is that the actual value of `j2me_enabled` is already being set correctly.) 

Now, it may very well be something you want to cache, so one strategy is to invalidate the cache every time a new build is added. To invalidate the cache I believe you can run `CommCareBuild.j2me_enabled_builds.clear(CommCareBuild)` but you'll probably want to double check that from `./manage.py shell`.

mkangia commented:
thanks @dannyroberts. I was not sure if adding a new build with a different build number but same version adds a new entry or updates the existing with that version, so if it was an update, an explicit set for j2me_enabled field was needed even though it was the default value.
@phillipm : Can you help confirm if its a caching issue? The cache is maintained for 5 minutes for available j2me supported build numbers. So even when you would do the subsequent submit, you should see the build with expected j2me_enabled value after 5 minutes(~7-8 for buffer) and not instantly. If thats the case i can just expire the cache like danny suggested.

dannyroberts commented:
@mkangia ah see what you mean. Two points to clarify:
- when a new build comes in, it just makes a new build document—it doesn't update any previous build document that has the same version
- whenever you're instantiating a new document instance (using `cls(...)` like in this line), when you call `.save()` on it, it'll make a **new** document, even if some of the fields are the same. So, there's no way that line could be updating a different document.
Hope that helps clarify those questions!

mkangia commented:
@thanks @dannyroberts. That was also my naive understanding at first but just was not sure. Curious now how we have just unique numbers for the versions on the edit_menu for builds, i guess it just picks the latest one. Will have a better look at it.

dannyroberts commented:
It pulls out unique versions here: https://github.com/dimagi/commcare-hq/blob/81fc1aa420e57e19e966d528a741299916ea2e76/corehq/apps/builds/utils.py#L14-14

mkangia commented:
Aha! Interesting. thanks @dannyroberts

mkangia committed:
Do not explicitly set j2me_enabled True since thats already the default set value. Expire cache on adding a new build using Zip file as it is j2me supported

mkangia commented:
@dannyroberts : Changes done for expiring cache. Please review.

dannyroberts commented:
Hey, this all looks good. Any further QA to be done? Once you're ready you can remove the "awaiting QA" label and I can merge it!

mkangia commented:
@dannyroberts I am trying to test [this](https://github.com/dimagi/commcare-hq/pull/12411/files#diff-a65d3e32d1001ad60472a36e6d006b88R99) on my machine though this was just added for an unexpected case. The build version does not go back to being non  j2me-enabled in case if resubmit without artifacts. So just need to confirm that.
Also, @phillipm can do a final test on staging and share his feedback.

dannyroberts commented:
👍

mkangia committed:
Pick the latest version build when checking for j2me_enabled builds

mkangia commented:
@dannyroberts : Please review [fix](https://github.com/dimagi/commcare-hq/pull/12411/commits/52c1a0310d5a0af9bb66393c1a35a3acfe06414b) to pick the latest build for version when deciding on j2me support. This should fix issue when creating a non-j2me build for same version post a j2me supported version. Tested it and works well for either cases.
Also, Do we need to sort this by build number or we can rely on their present order in couch db?

dannyroberts commented:
They'll come out in the order given by https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/builds/_design/views/all/map.js#L5-5, which is to say, sorted first by their `version` property and then by their `build_number` property (and then by time). By default the query is in ascending order, so the smaller build numbers would come first. `groupby` should preserve that order, so it looks (to me) like your code works correctly with that assumption.

mkangia commented:
@phillipm : Just a reminder on this. Awaiting final test.

phillipm commented:
Works perfectly! 👍

mkangia commented:
Awesome @phillipm.
@dannyroberts : Over to you now.

dannyroberts commented:
🎉

czue commented:
woop!

mkangia commented:
✌️ :metal: First feature implementation!!
